eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,327 @@
+		// This is a workround to mitigate the race conditions.
+		logger.Info("starting bitbucket hooks manager daemon")
+		ticker := time.NewTicker(60 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Info("exiting bitbucket hooks manager daemon")
+				return
+			case <-ticker.C:
+				err := router.CreateBitbucketWebhook(ctx, bitbucketConfig)
+
+				if err != nil {
+					logger.Errorw("failed to create/update Bitbucket webhook", zap.Error(err))
+				}
+			}
+		}
+	}()
+
+	return webhook.ManageRoute(ctx, router, controller, dispatch)
+}
+
+func (router *Router) CreateBitbucketWebhook(ctx context.Context, bitbucketConfig *bitbucketv1.Configuration) error {
+	bitbucketserverEventSource := router.bitbucketserverEventSource
+	route := router.route
+
+	logger := route.Logger.With(
+		logging.LabelEndpoint, route.Context.Endpoint,
+		logging.LabelPort, route.Context.Port,
+		logging.LabelHTTPMethod, route.Context.Method,
+		"project-key", bitbucketserverEventSource.ProjectKey,
+		"repository-slug", bitbucketserverEventSource.RepositorySlug,
+		"base-url", bitbucketserverEventSource.BitbucketServerBaseURL,
+	)
+
+	bitbucketClient := bitbucketv1.NewAPIClient(ctx, bitbucketConfig)
+
+	formattedURL := common.FormattedURL(bitbucketserverEventSource.Webhook.URL, bitbucketserverEventSource.Webhook.Endpoint)
+
+	apiResponse, err := bitbucketClient.DefaultApi.FindWebhooks(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to list existing hooks to check for duplicates for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	hooks, err := bitbucketv1.GetWebhooksResponse(apiResponse)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert the list of webhooks for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	var existingHook bitbucketv1.Webhook
+	isAlreadyExists := false
+
+	for _, hook := range hooks {
+		if hook.Url == formattedURL {
+			isAlreadyExists = true
+			existingHook = hook
+			router.hookID = hook.ID
+			break
+		}
+	}
+
+	logger.Info("retrieving the webhook secret...")
+	webhookSecret, err := common.GetSecretFromVolume(bitbucketserverEventSource.WebhookSecret)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver webhook secret. err: %+v", err)
+	}
+
+	newHook := bitbucketv1.Webhook{
+		Name:          "Argo Events",
+		Url:           formattedURL,
+		Active:        true,
+		Events:        bitbucketserverEventSource.Events,
+		Configuration: bitbucketv1.WebhookConfiguration{Secret: webhookSecret},
+	}
+
+	localVarPostBody, err := json.Marshal(newHook)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal new webhook to JSON")
+	}
+
+	var createdHook *bitbucketv1.Webhook
```

Comment:
Minor: move it into `if` block below.

Reply:
Thanks. That should've prevented me from making the previous mistake.